src/utils/parser/getScopes.js

```diff
@@ -0,0 +1,280 @@
+// @flow
+
+import { traverseAst } from "./utils/ast";
+import { getSource } from "./sources";
+
+import type { Source, SourceId, Location, SourceScope } from "debugger-html";
+import type { NodePath, Node, Location as BabelLocation } from "babel-traverse";
+
+export type ScopeDeclaration = {
+  type: string,
+  start: Location,
+  end: Location,
+  bindings: any
+};
+
+export type ScopeDeclarations = ScopeDeclaration[];
+
+type GeneratedScope = {
+  start: Location,
+  end: Location,
+  type: string,
+  bindings: any,
+  children: ?(GeneratedScope[])
+};
+
+type TempScope = {
+  type: string,
+  parent: TempScope | null,
+  scopes: Array<TempScope>,
+  loc: BabelLocation,
+  names: any
+};
+
+function _createTempScope(
+  type: string,
+  parent: TempScope | null,
+  loc: BabelLocation
+): TempScope {
+  let result = {
+    type,
+    parent,
+    scopes: [],
+    loc: loc,
+    names: Object.create(null)
+  };
+  if (parent) {
+    parent.scopes.push(result);
+  }
+  return result;
+}
+
+function isNode(node?: Node, type: string): boolean {
+  return node ? node.type === type : false;
+}
+
+function _getFunctionScope(scope: TempScope): TempScope {
+  let s = scope;
+  while (s.type !== "Function" && s.type !== "Script") {
+    if (!s.parent) {
+      return s;
+    }
+    s = s.parent;
+  }
+  return s;
+}
+
+function _fromBabelLocation(
+  location: BabelLocation,
+  sourceId: SourceId
+): Location {
+  return {
+    sourceId,
+    line: location.line,
+    column: location.column
+  };
+}
+
+function _parseDeclarator(
+  declaratorId: Node,
+  targetScope: TempScope,
+  kind: string
+) {
+  if (isNode(declaratorId, "Identifier")) {
+    targetScope.names[declaratorId.name] = { kind, refs: [] };
+    return;
+  } else if (isNode(declaratorId, "ObjectPattern")) {
+    declaratorId.properties.forEach(prop => {
+      _parseDeclarator(prop.key, targetScope, kind);
+    });
+  } else if (isNode(declaratorId, "ArrayPattern")) {
+    declaratorId.elements.forEach(item => {
+      _parseDeclarator(item, targetScope, kind);
+    });
+  } else if (isNode(declaratorId, "RestElement")) {
+    _parseDeclarator(declaratorId.argument, targetScope, kind);
+  }
+}
+
+function _recordIdentifierPosition(
+  scope: TempScope,
+  name: string,
+  loc: BabelLocation
+) {
+  for (let s = scope; s; s = s.parent) {
+    if (name in s.names) {
+      s.names[name].refs.push(loc.start);
+      break;
+    }
+  }
+}
+
+function _toGeneratedScopes(
+  children: TempScope[],
+  sourceId: SourceId
+): ?(GeneratedScope[]) {
+  if (!children || children.length === 0) {
+    return undefined;
+  }
+  return children.map(scope => {
+    let bindings = Object.create(null);
+    Object.keys(scope.names).forEach(n => {
+      switch (scope.names[n].kind) {
+        case "var":
+        case "let":
+        case "const":
+        case "param":
+          bindings[n] = scope.names[n].refs.map(location => {
```

Comment:
i think i have reviewed a very similar piece of code here: https://github.com/devtools-html/debugger.html/pull/3852/files/e0f9e89e7ee904fbd1d1a17a0d801e8f7a1965dc#diff-dcf4c6f392912b09dabbbf7f6d2436b4

rather than copying this over, maybe check there first -- the same comment apply here